homeassistant/components/modbus/base_platform.py

```diff
@@ -161,6 +164,10 @@ def __init__(self, hub: ModbusHub, config: dict) -> None:
         self._scale = config[CONF_SCALE]
         self._offset = config[CONF_OFFSET]
         self._count = config[CONF_COUNT]
+        if self._nan_value is not None:
+            self._nan_value = struct.unpack(
+                self._structure, bytes.fromhex(self._nan_value[2:])
```

Comment:
yes, otherwise pythons `int('hex_string')` yields the integer representation of `'hex_string'`regardless of the modbus datatype, which can be `uint16`, `int16`, `uint32`, `int32`, `float32`,... and all as either little endian or big endian. 
The `'hex_string'` NaN value has to be decoded in the same way as the actual incoming modbus sensor value is decoded.

Reply:
see above this breaks with custom formats.

Reply:
Couldn't find a good solution yet